File: SEALNETTest/BigPolyWrapper.cs
using System.IO;
using Microsoft.Research.SEAL;
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace SEALNETTest
{
    [TestClass]
    public unsafe class BigPolyWrapper
    {
        [TestMethod]
        public void EmptyBigPolyNET()
        {
            var poly = new BigPoly();
            Assert.AreEqual(0, poly.CoeffCount);
            Assert.AreEqual(0, poly.CoeffBitCount);
            Assert.AreEqual(0, poly.UInt64Count);
            Assert.IsTrue(null == poly.Pointer);
            Assert.AreEqual(0, poly.GetSignificantCoeffCount());
            Assert.IsTrue(poly.IsZero);
            Assert.IsFalse(poly.IsAlias);
            Assert.AreEqual("0", poly.ToString());
            poly.SetZero();

            var poly2 = new BigPoly();
            Assert.IsTrue(poly.Equals(poly2));

            poly.Resize(1, 1);
            Assert.IsTrue(null != poly.Pointer);
            Assert.IsFalse(poly.IsAlias);

            poly.Resize(0, 0);
            Assert.IsTrue(null == poly.Pointer);
            Assert.IsFalse(poly.IsAlias);
        }

        [TestMethod]
        public void BigPoly3Coeff100BitsNET()
        {
            var poly = new BigPoly(3, 100);
            Assert.AreEqual(3, poly.CoeffCount);
            Assert.AreEqual(100, poly.CoeffBitCount);
            Assert.IsTrue(null != poly.Pointer);
            Assert.AreEqual(6, poly.UInt64Count);
            Assert.AreEqual(0, poly.GetSignificantCoeffCount());
            Assert.AreEqual("0", poly.ToString());
            Assert.IsTrue(poly.IsZero);
            Assert.AreEqual(0UL, poly.Pointer[0]);
            Assert.AreEqual(0UL, poly.Pointer[1]);
            Assert.AreEqual(0UL, poly.Pointer[2]);
            Assert.AreEqual(0UL, poly.Pointer[3]);
            Assert.AreEqual(0UL, poly.Pointer[4]);
            Assert.AreEqual(0UL, poly.Pointer[5]);

            poly[0].Set("1234");
            Assert.AreEqual("1234", poly[0].ToString());
            Assert.AreEqual(1, poly.GetSignificantCoeffCount());
            Assert.AreEqual("1234", poly.ToString());
            Assert.IsFalse(poly.IsZero);
            Assert.AreEqual(0x1234UL, poly.Pointer[0]);
            Assert.AreEqual(0UL, poly.Pointer[1]);
            Assert.AreEqual(0UL, poly.Pointer[2]);
            Assert.AreEqual(0UL, poly.Pointer[3]);
            Assert.AreEqual(0UL, poly.Pointer[4]);
            Assert.AreEqual(0UL, poly.Pointer[5]);

            poly[2].Set("3211234567812345678");
            Assert.AreEqual("3211234567812345678", poly[2].ToString());
            Assert.AreEqual(3, poly.GetSignificantCoeffCount());
            Assert.AreEqual("3211234567812345678x^2 + 1234", poly.ToString());
            Assert.IsFalse(poly.IsZero);
            Assert.AreEqual(0x1234UL, poly.Pointer[0]);
            Assert.AreEqual(0UL, poly.Pointer[1]);
            Assert.AreEqual(0UL, poly.Pointer[2]);
            Assert.AreEqual(0UL, poly.Pointer[3]);
            Assert.AreEqual(0x1234567812345678UL, poly.Pointer[4]);
            Assert.AreEqual(0x321UL, poly.Pointer[5]);

            poly.SetZero();
            Assert.IsTrue(poly.IsZero);
            Assert.AreEqual("0", poly.ToString());
            Assert.AreEqual(0UL, poly.Pointer[0]);
            Assert.AreEqual(0UL, poly.Pointer[1]);
            Assert.AreEqual(0UL, poly.Pointer[2]);
            Assert.AreEqual(0UL, poly.Pointer[3]);
            Assert.AreEqual(0UL, poly.Pointer[4]);
            Assert.AreEqual(0UL, poly.Pointer[5]);

            poly[0].Set("123");
            poly[1].Set("456");
            poly[2].Set("3211234567812345678");
            Assert.AreEqual("3211234567812345678x^2 + 456x^1 + 123", poly.ToString());
            Assert.AreEqual(3, poly.GetSignificantCoeffCount());
            Assert.AreEqual(0x123UL, poly.Pointer[0]);
            Assert.AreEqual(0UL, poly.Pointer[1]);
            Assert.AreEqual(0x456UL, poly.Pointer[2]);
            Assert.AreEqual(0UL, poly.Pointer[3]);
            Assert.AreEqual(0x1234567812345678UL, poly.Pointer[4]);
            Assert.AreEqual(0x321UL, poly.Pointer[5]);

            var poly2 = new BigPoly(3, 64);
            poly2[0].Set(1);
            poly2[1].Set(2);
            poly2[2].Set(3);
            Assert.IsFalse(poly.Equals(poly2));
            Assert.IsFalse(poly2.Equals(poly));
            Assert.AreNotEqual(poly.GetHashCode(), poly2.GetHashCode());
            poly.Set(poly2);
            Assert.IsTrue(poly.Equals(poly2));
            Assert.IsTrue(poly2.Equals(poly));
            Assert.AreEqual(poly.GetHashCode(), poly2.GetHashCode());
            Assert.AreEqual("3x^2 + 2x^1 + 1", poly.ToString());
            Assert.AreEqual(1UL, poly.Pointer[0]);
            Assert.AreEqual(0UL, poly.Pointer[1]);
            Assert.AreEqual(2UL, poly.Pointer[2]);
            Assert.AreEqual(0UL, poly.Pointer[3]);
            Assert.AreEqual(3UL, poly.Pointer[4]);
            Assert.AreEqual(0UL, poly.Pointer[5]);

            poly[0].Set("123");
            poly[1].Set("456");
            poly[2].Set("3211234567812345678");
            poly.Resize(3, 32);
            Assert.AreEqual(3, poly.CoeffCount);
            Assert.AreEqual(32, poly.CoeffBitCount);
            Assert.AreEqual(3, poly.UInt64Count);
            Assert.AreEqual("12345678x^2 + 456x^1 + 123", poly.ToString());
            Assert.AreEqual(0x123UL, poly.Pointer[0]);
            Assert.AreEqual(0x456UL, poly.Pointer[1]);
            Assert.AreEqual(0x12345678UL, poly.Pointer[2]);

            poly.Resize(4, 100);
            Assert.AreEqual(4, poly.CoeffCount);
            Assert.AreEqual(100, poly.CoeffBitCount);
            Assert.AreEqual("12345678x^2 + 456x^1 + 123", poly.ToString());
            Assert.AreEqual(0x123UL, poly.Pointer[0]);
            Assert.AreEqual(0UL, poly.Pointer[1]);
            Assert.AreEqual(0x456UL, poly.Pointer[2]);
            Assert.AreEqual(0UL, poly.Pointer[3]);
            Assert.AreEqual(0x12345678UL, poly.Pointer[4]);
            Assert.AreEqual(0UL, poly.Pointer[5]);
            Assert.AreEqual(0UL, poly.Pointer[6]);
            Assert.AreEqual(0UL, poly.Pointer[7]);

            poly.Resize(0, 0);
            Assert.AreEqual(0, poly.CoeffCount);
            Assert.AreEqual(0, poly.CoeffBitCount);
            Assert.AreEqual(0, poly.UInt64Count);
            Assert.IsTrue(null == poly.Pointer);
            Assert.AreEqual(0, poly.GetSignificantCoeffCount());
            Assert.IsTrue(poly.IsZero);
            Assert.IsFalse(poly.IsAlias);
            Assert.AreEqual("0", poly.ToString());
        }

        [TestMethod]
        public void BigPolyFromStringNET()
        {
            var poly = new BigPoly("9x^2 + 2");
            Assert.AreEqual("9x^2 + 2", poly.ToString());
            Assert.AreEqual(3, poly.CoeffCount);
            Assert.AreEqual(4, poly.CoeffBitCount);

            poly.Resize(0, 0);
            poly.Set("0");
            Assert.AreEqual("0", poly.ToString());
            Assert.AreEqual(0, poly.CoeffCount);
            Assert.AreEqual(0, poly.CoeffBitCount);

            poly.Set("1");
            Assert.AreEqual("1", poly.ToString());
            Assert.AreEqual(1, poly.CoeffCount);
            Assert.AreEqual(1, poly.CoeffBitCount);

            poly.Set("123");
            Assert.AreEqual("123", poly.ToString());
            Assert.AreEqual(1, poly.CoeffCount);
            Assert.AreEqual(9, poly.CoeffBitCount);

            poly.Set("23x^3");
            Assert.AreEqual("23x^3", poly.ToString());
            Assert.AreEqual(4, poly.CoeffCount);
            Assert.AreEqual(9, poly.CoeffBitCount);

            poly.Resize(0, 0);
            poly.Set("23x^3");
            Assert.AreEqual("23x^3", poly.ToString());
            Assert.AreEqual(4, poly.CoeffCount);
            Assert.AreEqual(6, poly.CoeffBitCount);

            poly.Set("3x^2 + 2x^1");
            Assert.AreEqual("3x^2 + 2x^1", poly.ToString());
            Assert.AreEqual(4, poly.CoeffCount);
            Assert.AreEqual(6, poly.CoeffBitCount);

            poly.Set("1x^2 + 2x^1 + 3");
            Assert.AreEqual("1x^2 + 2x^1 + 3", poly.ToString());
            Assert.AreEqual(4, poly.CoeffCount);
            Assert.AreEqual(6, poly.CoeffBitCount);

            poly.Set("4x^1");
            Assert.AreEqual("4x^1", poly.ToString());
            Assert.AreEqual(4, poly.CoeffCount);
            Assert.AreEqual(6, poly.CoeffBitCount);
        }

        [TestMethod]
        public void SaveLoadPolyNET()
        {
            var stream = new MemoryStream();

            var poly = new BigPoly();
            var poly2 = new BigPoly(1, 32);
            poly2[0].Set(1);
            stream.Seek(0, SeekOrigin.Begin);
            poly.Save(stream);
            stream.Seek(0, SeekOrigin.Begin);
            poly2.Load(stream);
            Assert.AreEqual(poly, poly2);

            poly.Resize(3, 96);
            poly[0].Set(2);
            poly[1].Set(3);
            poly[2].Set(4);
            stream.Seek(0, SeekOrigin.Begin);
            poly.Save(stream);
            stream.Seek(0, SeekOrigin.Begin);
            poly2.Load(stream);
            Assert.AreEqual(poly, poly2);

            poly[0].Set("FFFFFFFFFFFFFFFFFF");
            stream.Seek(0, SeekOrigin.Begin);
            poly.Save(stream);
            stream.Seek(0, SeekOrigin.Begin);
            poly2.Load(stream);
            Assert.AreEqual(poly, poly2);

            poly.Resize(2, 32);
            poly[0].Set(1);
            stream.Seek(0, SeekOrigin.Begin);
            poly.Save(stream);
            stream.Seek(0, SeekOrigin.Begin);
            poly2.Load(stream);
            Assert.AreEqual(poly, poly2);
        }

        [TestMethod]
        public void DuplicateToNET()
        {
            var original = new BigPoly(123, 456);
            original[0].Set(1);
            original[1].Set(2);
            original[2].Set(3);
            original[3].Set(4);
            original[4].Set(5);
            original[122].Set(123);

            BigPoly target = new BigPoly();

            original.DuplicateTo(target);
            Assert.AreEqual(target.CoeffCount, original.CoeffCount);
            Assert.AreEqual(target.CoeffBitCount, original.CoeffBitCount);
            Assert.IsTrue(target.Equals(original));
        }

        [TestMethod]
        public void DuplicateFromNET()
        {
            var original = new BigPoly(123, 456);
            original[0].Set(1);
            original[1].Set(2);
            original[2].Set(3);
            original[3].Set(4);
            original[4].Set(5);
            original[122].Set(123);

            BigPoly target = new BigPoly();

            target.DuplicateFrom(original);
            Assert.AreEqual(target.CoeffCount, original.CoeffCount);
            Assert.AreEqual(target.CoeffBitCount, original.CoeffBitCount);
            Assert.IsTrue(target.Equals(original));
        }
    }
}

File: SEALNET/BigPolyWrapper.cpp
#include <msclr\marshal_cppstd.h>
#include <stdexcept>
#include <algorithm>
#include "Common.h"
#include "util/common.h"
#include "util/uintcore.h"
#include "util/polycore.h"
#include "BigPolyWrapper.h"
#include "BigUIntWrapper.h"

using namespace System;
using namespace System::IO;
using namespace msclr::interop;
using namespace std;
using namespace seal::util;

namespace Microsoft
{
    namespace Research
    {
        namespace SEAL
        {
            BigPoly::BigPoly() : bigpoly_(nullptr), owned_(true)
            {
                try
                {
                    bigpoly_ = new seal::BigPoly();
                }
                catch (const exception &e)
                {
                    HandleException(&e);
                }
                catch (...)
                {
                    HandleException(nullptr);
                }
            }

            BigPoly::BigPoly(int coeffCount, int coeffBitCount) : bigpoly_(nullptr), owned_(true)
            {
                try
                {
                    bigpoly_ = new seal::BigPoly(coeffCount, coeffBitCount);
                }
                catch (const exception &e)
                {
                    HandleException(&e);
                }
                catch (...)
                {
                    HandleException(nullptr);
                }
            }

            BigPoly::BigPoly(String ^hexPoly) : bigpoly_(nullptr), owned_(true)
            {
                if (hexPoly == nullptr)
                {
                    throw gcnew ArgumentNullException("hexPoly cannot be null");
                }
                marshal_context ^context = gcnew marshal_context();
                try
                {
                    bigpoly_ = new seal::BigPoly(context->marshal_as<const char*>(hexPoly));
                    GC::KeepAlive(hexPoly);
                }
                catch (const exception &e)
                {
                    HandleException(&e);
                }
                catch (...)
                {
                    HandleException(nullptr);
                }
                finally
                {
                    delete context;
                }
            }

            BigPoly::BigPoly(int coeffCount, int coeffBitCount, String ^hexPoly) : bigpoly_(nullptr), owned_(true)
            {
                if (hexPoly == nullptr)
                {
                    throw gcnew ArgumentNullException("hexPoly cannot be null");
                }
                marshal_context ^context = gcnew marshal_context();
                try
                {
                    bigpoly_ = new seal::BigPoly(coeffCount, coeffBitCount, context->marshal_as<const char*>(hexPoly));
                }
                catch (const exception &e)
                {
                    HandleException(&e);
                }
                catch (...)
                {
                    HandleException(nullptr);
                }
                finally
                {
                    delete context;
                }
            }

            BigPoly::BigPoly(BigPoly ^copy) : bigpoly_(nullptr), owned_(true)
            {
                if (copy == nullptr)
                {
                    throw gcnew ArgumentNullException("copy cannot be null");
                }
                try
                {
                    bigpoly_ = new seal::BigPoly(copy->GetPolynomial());
                    GC::KeepAlive(copy);
                }
                catch (const exception &e)
                {
                    HandleException(&e);
                }
                catch (...)
                {
                    HandleException(nullptr);
                }
            }

            bool BigPoly::IsAlias::get()
            {
                if (bigpoly_ == nullptr)
                {
                    throw gcnew ObjectDisposedException("BigPoly is disposed");
                }
                return bigpoly_->is_alias();
            }

            int BigPoly::CoeffCount::get()
            {
                if (bigpoly_ == nullptr)
                {
                    throw gcnew ObjectDisposedException("BigPoly is disposed");
                }
                return bigpoly_->coeff_count();
            }

            int BigPoly::CoeffBitCount::get()
            {
                if (bigpoly_ == nullptr)
                {
                    throw gcnew ObjectDisposedException("BigPoly is disposed");
                }
                return bigpoly_->coeff_bit_count();
            }

            int BigPoly::UInt64Count::get()
            {
                if (bigpoly_ == nullptr)
                {
                    throw gcnew ObjectDisposedException("BigPoly is disposed");
                }
                return bigpoly_->uint64_count();
            }

            int BigPoly::CoeffUInt64Count::get()
            {
                if (bigpoly_ == nullptr)
                {
                    throw gcnew ObjectDisposedException("BigPoly is disposed");
                }
                return bigpoly_->coeff_uint64_count();
            }

            UInt64 *BigPoly::Pointer::get()
            {
                if (bigpoly_ == nullptr)
                {
                    throw gcnew ObjectDisposedException("BigPoly is disposed");
                }
                return bigpoly_->pointer();
            }

            bool BigPoly::IsZero::get()
            {
                if (bigpoly_ == nullptr)
                {
                    throw gcnew ObjectDisposedException("BigPoly is disposed");
                }
                return bigpoly_->is_zero();
            }

            BigUInt ^BigPoly::default::get(int index)
            {
                if (bigpoly_ == nullptr)
                {
                    throw gcnew ObjectDisposedException("BigPoly is disposed");
                }
                try
                {
                    return gcnew BigUInt(&bigpoly_->operator[](index));
                }
                catch (const exception &e)
                {
                    HandleException(&e);
                }
                catch (...)
                {
                    HandleException(nullptr);
                }
                throw gcnew Exception("Unexpected exception");
            }

            int BigPoly::GetSignificantCoeffCount()
            {
                if (bigpoly_ == nullptr)
                {
                    throw gcnew ObjectDisposedException("BigPoly is disposed");
                }
                return bigpoly_->significant_coeff_count();
            }

            int BigPoly::GetSignificantCoeffBitCount()
            {
                if (bigpoly_ == nullptr)
                {
                    throw gcnew ObjectDisposedException("BigPoly is disposed");
                }
                return bigpoly_->significant_coeff_bit_count();
            }

            void BigPoly::Set(BigPoly ^assign)
            {
                if (bigpoly_ == nullptr)
                {
                    throw gcnew ObjectDisposedException("BigPoly is disposed");
                }
                if (assign == nullptr)
                {
                    throw gcnew ArgumentNullException("assign cannot be null");
                }
                try
                {
                    *bigpoly_ = assign->GetPolynomial();
                    GC::KeepAlive(assign);
                }
                catch (const exception &e)
                {
                    HandleException(&e);
                }
                catch (...)
                {
                    HandleException(nullptr);
                }
            }

            void BigPoly::Set(String ^assign)
            {
                if (bigpoly_ == nullptr)
                {
                    throw gcnew ObjectDisposedException("BigPoly is disposed");
                }
                if (assign == nullptr)
                {
                    throw gcnew ArgumentNullException("assign cannot be null");
                }
                marshal_context ^context = gcnew marshal_context();
                try
                {
                    *bigpoly_ = context->marshal_as<const char*>(assign);
                    GC::KeepAlive(assign);
                }
                catch (const exception &e)
                {
                    HandleException(&e);
                }
                catch (...)
                {
                    HandleException(nullptr);
                }
                finally
                {
                    delete context;
                }
            }

            void BigPoly::SetZero()
            {
                if (bigpoly_ == nullptr)
                {
                    throw gcnew ObjectDisposedException("BigPoly is disposed");
                }
                bigpoly_->set_zero();
            }

            void BigPoly::SetZero(int startCoeff)
            {
                if (bigpoly_ == nullptr)
                {
                    throw gcnew ObjectDisposedException("BigPoly is disposed");
                }
                bigpoly_->set_zero(startCoeff);
            }

            void BigPoly::SetZero(int startCoeff, int coeffCount)
            {
                if (bigpoly_ == nullptr)
                {
                    throw gcnew ObjectDisposedException("BigPoly is disposed");
                }
                bigpoly_->set_zero(startCoeff, coeffCount);
            }

            void BigPoly::Save(Stream ^stream)
            {
                if (bigpoly_ == nullptr)
                {
                    throw gcnew ObjectDisposedException("BigPoly is disposed");
                }
                if (stream == nullptr)
                {
                    throw gcnew ArgumentNullException("stream cannot be null");
                }
                try
                {
                    int32_t coeff_count32 = static_cast<int32_t>(bigpoly_->coeff_count());
                    Write(stream, reinterpret_cast<const char*>(&coeff_count32), sizeof(int32_t));
                    int32_t coeff_bit_count32 = static_cast<int32_t>(bigpoly_->coeff_bit_count());
                    Write(stream, reinterpret_cast<const char*>(&coeff_bit_count32), sizeof(int32_t));
                    int coeff_uint64_count = divide_round_up(bigpoly_->coeff_bit_count(), bits_per_uint64);
                    Write(stream, reinterpret_cast<const char*>(bigpoly_->pointer()), bigpoly_->coeff_count() * coeff_uint64_count * bytes_per_uint64);
                
                    GC::KeepAlive(stream);
                }
                catch (const exception &e)
                {
                    HandleException(&e);
                }
                catch (Exception^)
                {
                    throw;
                }
                catch (...)
                {
                    HandleException(nullptr);
                }
            }

            void BigPoly::Load(Stream ^stream)
            {
                if (bigpoly_ == nullptr)
                {
                    throw gcnew ObjectDisposedException("BigPoly is disposed");
                }
                if (stream == nullptr)
                {
                    throw gcnew ArgumentNullException("stream cannot be null");
                }
                try
                {
                    int32_t read_coeff_count = 0;
                    Read(stream, reinterpret_cast<char*>(&read_coeff_count), sizeof(int32_t));
                    int32_t read_coeff_bit_count = 0;
                    Read(stream, reinterpret_cast<char*>(&read_coeff_bit_count), sizeof(int32_t));
                    if (read_coeff_count > bigpoly_->coeff_count() || read_coeff_bit_count > bigpoly_->coeff_bit_count())
                    {
                        // Size is too large to currently fit, so resize.
                        bigpoly_->resize(std::max(read_coeff_count, bigpoly_->coeff_count()), std::max(read_coeff_bit_count, bigpoly_->coeff_bit_count()));
                    }
                    int read_coeff_uint64_count = divide_round_up(read_coeff_bit_count, bits_per_uint64);
                    int coeff_uint64_count = divide_round_up(bigpoly_->coeff_bit_count(), bits_per_uint64);
                    if (read_coeff_uint64_count == coeff_uint64_count)
                    {
                        Read(stream, reinterpret_cast<char*>(bigpoly_->pointer()), read_coeff_count * coeff_uint64_count * bytes_per_uint64);
                    }
                    else
                    {
                        // Coefficients are different sizes, so read one at a time.
                        uint64_t *coeff_ptr = bigpoly_->pointer();
                        for (int i = 0; i < read_coeff_count; ++i)
                        {
                            Read(stream, reinterpret_cast<char*>(coeff_ptr), read_coeff_uint64_count * bytes_per_uint64);
                            set_zero_uint(coeff_uint64_count - read_coeff_uint64_count, coeff_ptr + read_coeff_uint64_count);
                            coeff_ptr += coeff_uint64_count;
                        }
                    }

                    // Zero any remaining coefficients.
                    if (bigpoly_->coeff_count() > read_coeff_count)
                    {
                        set_zero_poly(bigpoly_->coeff_count() - read_coeff_count, coeff_uint64_count, bigpoly_->pointer() + read_coeff_count * coeff_uint64_count);
                    }

                    GC::KeepAlive(stream);
                }
                catch (const exception &e)
                {
                    HandleException(&e);
                }
                catch (Exception^)
                {
                    throw;
                }
                catch (...)
                {
                    HandleException(nullptr);
                }
            }

            void BigPoly::Resize(int coeffCount, int coeffBitCount)
            {
                if (bigpoly_ == nullptr)
                {
                    throw gcnew ObjectDisposedException("BigPoly is disposed");
                }
                try
                {
                    bigpoly_->resize(coeffCount, coeffBitCount);
                }
                catch (const exception &e)
                {
                    HandleException(&e);
                }
                catch (...)
                {
                    HandleException(nullptr);
                }
            }

            bool BigPoly::Equals(BigPoly ^compare)
            {
                if (bigpoly_ == nullptr)
                {
                    throw gcnew ObjectDisposedException("BigPoly is disposed");
                }
                if (compare == nullptr)
                {
                    return false;
                }
                auto result = *bigpoly_ == compare->GetPolynomial();
                GC::KeepAlive(compare);
                return result;
            }

            bool BigPoly::Equals(Object ^compare)
            {
                if (bigpoly_ == nullptr)
                {
                    throw gcnew ObjectDisposedException("BigPoly is disposed");
                }
                BigPoly ^comparepoly = dynamic_cast<BigPoly^>(compare);
                if (comparepoly != nullptr)
                {
                    return Equals(comparepoly);
                }
                return false;
            }

            String ^BigPoly::ToString()
            {
                if (bigpoly_ == nullptr)
                {
                    throw gcnew ObjectDisposedException("BigPoly is disposed");
                }
                return marshal_as<String^>(bigpoly_->to_string());
            }

            int BigPoly::GetHashCode()
            {
                if (bigpoly_ == nullptr)
                {
                    throw gcnew ObjectDisposedException("BigPoly is disposed");
                }
                return ComputeArrayHashCode(bigpoly_->pointer(), bigpoly_->uint64_count());
            }

            seal::BigPoly &BigPoly::GetPolynomial()
            {
                if (bigpoly_ == nullptr)
                {
                    throw gcnew ObjectDisposedException("BigPoly is disposed");
                }
                return *bigpoly_;
            }

            BigPoly::~BigPoly()
            {
                this->!BigPoly();
            }

            BigPoly::!BigPoly()
            {
                if (bigpoly_ != nullptr)
                {
                    if (owned_)
                    {
                        delete bigpoly_;
                    }
                    bigpoly_ = nullptr;
                }
            }

            BigPoly::BigPoly(const seal::BigPoly &value) : bigpoly_(nullptr), owned_(true)
            {
                try
                {
                    bigpoly_ = new seal::BigPoly(value);
                }
                catch (const exception &e)
                {
                    HandleException(&e);
                }
                catch (...)
                {
                    HandleException(nullptr);
                }
            }

            BigPoly::BigPoly(seal::BigPoly *value) : bigpoly_(value), owned_(false)
            {
            }

            void BigPoly::DuplicateTo(BigPoly ^destination)
            {
                if (bigpoly_ == nullptr)
                {
                    throw gcnew ObjectDisposedException("BigPoly is disposed");
                }
                if (destination == nullptr)
                {
                    throw gcnew ArgumentNullException("destination cannot be null");
                }
                try
                {
                    bigpoly_->duplicate_to(*destination->bigpoly_);
                    GC::KeepAlive(destination);
                }
                catch (const exception &e)
                {
                    HandleException(&e);
                }
                catch (...)
                {
                    HandleException(nullptr);
                }
            }

            void BigPoly::DuplicateFrom(BigPoly ^value)
            {
                if (bigpoly_ == nullptr)
                {
                    throw gcnew ObjectDisposedException("BigPoly is disposed");
                }
                if (value == nullptr)
                {
                    throw gcnew ArgumentNullException("value cannot be null");
                }
                try
                {
                    bigpoly_->duplicate_from(*value->bigpoly_);
                    GC::KeepAlive(value);
                }
                catch (const exception &e)
                {
                    HandleException(&e);
                }
                catch (...)
                {
                    HandleException(nullptr);
                }
            }
        }
    }
}